text/2021-12-29-openapi-to-typescript-api-client-and-forms.md

```diff
@@ -0,0 +1,458 @@
+
+Then we can use create a new package to handle the generation:
+
+```sh
+mkdir -p ui/packages/openapi
+cd ui/package/openapi
+yarn init
+```
+
+Install dependencies and add the `generate` script:
+
+```json
+{
+  "scripts": {
+    "generate": "export TS_POST_PROCESS_FILE='../../node_modules/.bin/prettier --write'; openapi-generator-cli generate -c openapiconfig.json -i ../../../pkg/dashboard/swaggerdocs/swagger.yaml -g typescript-axios -o ../../app/src/openapi --enable-post-process-file"
+  }
+}
+```
+
+This script will output the files into `ui/app/src/openapi` directory directly to
+to avoid additional compilation.
+
+### Use TypeScript Compiler API to Generate Forms
+
+To generate forms, we have think about these things before we write the code:
+
+- Construct the interface of a form field.
+- How to handle dependencies between fields?
+- How to find shared fields?
+- How to use the generated schemas to construct the interface?
+
+For the first item, a form field should normally contain at least:
+
+- type
+- label
+- value
+- description
+
+Convert to a JSON:
+
+```json
+{
+  "field: "text",
+  "label": "Name",
+  "value": "",
+  "helperText": "Fill your name",
+}
+```
+
+The types we currently have are:
+
+```ts
+type FieldType =
+  | "text"
+  | "textarea"
+  | "number"
+  | "select"
+  | "label"
+  | "autocomplete";
+```
+
+Most of them are inherited from the HTML Input types, except `label` and `autocomplete`.
+The `label` is represented as `string[]`. The `autocomplete` can be ignored for
+now because all generated fields won't use it.
+
+Done. Next we need to think about how to handle dependencies between fields.
+
+#### Dependencies between fields
+
+Mostly, we use an `action` field to distinguish exactly what we want a chaos to
+do with the injection. So I'll start from here, but how we can find different actions?
+The key problem is OpenAPI generator can't convert Go `type alias` to TS `enums`,
+for example:
+
+```go
+const (
+        Ec2Stop AWSChaosAction = "ec2-stop"
+        Ec2Restart AWSChaosAction = "ec2-restart"
+        DetachVolume AWSChaosAction = "detach-volume"
+)
+```
+
+We expect these can covert to:
+
+```ts
+enum AWSChaosAction {
+  Ec2Stop = "ec2-stop",
+  Ec2Restart = "ec2-restart",
+  DetachVolume = "detach-volume",
+}
+```
+
+But unfortunately at last we only have:
+
+```ts
+export interface V1alpha1AWSChaosSpec {
+  /**
+   * @type {string}
+   * @memberof V1alpha1AWSChaosSpec
+   */
+  action?: string;
+
+  //...
+}
+```
+
+This prevents us from using the TS compiler API to read the enum, but the solution
+is also simple, **since we can't get the key information through the code, we need
+to define it manually**.
+
+How about defining a json file? This is what I first thought, but @STRRL reminded
+me that **we can write it in the comments**. Yes, like `kubebuilder`'s marker, we
+can also define markers.
+
+So the following are all markers we will using:
+
+- ui:form:enum=action1;action2;action3
```

Comment:
Could we reuse `+kubebuilder:validation:Enum`, like `+kubebuilder:validation:Enum` is an alias for `ui:form:enum`.

Reply:
Makes sense! This saves a lot of work but requires a special note. I will update it soon.